File: 2025/12/03/structured-outputs-llms-reliable-json/typescript/src/parser.ts
/**
 * Extract JSON from raw LLM responses.
 */

export function extractJson(text: string): Record<string, any> | any[] {
  // Step 1: Trim whitespace
  text = text.trim();
  
  // Step 2: Remove markdown code blocks
  // Matches ```json ... ``` or ``` ... ```
  text = text.replace(/```(?:json)?\s*\n?(.*?)\n?```/gs, '$1');
  text = text.trim();
  
  // Step 3: Find JSON object/array
  // Look for first { or [
  let start = text.indexOf('{');
  if (start === -1) {
    start = text.indexOf('[');
  }
  if (start === -1) {
    throw new Error('No JSON found in response');
  }
  
  // Step 4: Find matching closing brace/bracket
  let depth = 0;
  let inString = false;
  let escapeNext = false;
  
  for (let i = start; i < text.length; i++) {
    const char = text[i];
    
    if (escapeNext) {
      escapeNext = false;
      continue;
    }
    
    if (char === '\\') {
      escapeNext = true;
      continue;
    }
    
    if (char === '"' && !escapeNext) {
      inString = !inString;
      continue;
    }
    
    if (inString) {
      continue;
    }
    
    if (char === '{' || char === '[') {
      depth++;
    } else if (char === '}' || char === ']') {
      depth--;
      if (depth === 0) {
        const jsonStr = text.substring(start, i + 1);
        try {
          return JSON.parse(jsonStr);
        } catch (e: any) {
          throw new Error(`Invalid JSON: ${e.message}`);
        }
      }
    }
  }
  
  throw new Error('Unclosed JSON structure');
}

export function extractJsonSimple(text: string): Record<string, any> | any[] | null {
  text = text.trim();
  
  // Try parsing the whole text first
  try {
    return JSON.parse(text);
  } catch {
    // Fall back to full extraction
    try {
      return extractJson(text);
    } catch {
      return null;
    }
  }
}

File: 2025/12/03/structured-outputs-llms-reliable-json/python/src/parser.py
"""Extract JSON from raw LLM responses."""
import json
import re
from typing import Any


def extract_json(text: str) -> dict | list:
    """
    Extract JSON from text that might contain markdown, explanations, or extra text.
    
    Args:
        text: Raw text that might contain JSON
    
    Returns:
        Parsed JSON object or array
    
    Raises:
        ValueError: If no valid JSON can be extracted
    """
    # Step 1: Trim whitespace
    text = text.strip()
    
    # Step 2: Remove markdown code blocks
    # Matches ```json ... ``` or ``` ... ```
    text = re.sub(r'```(?:json)?\s*\n?(.*?)\n?```', r'\1', text, flags=re.DOTALL)
    text = text.strip()
    
    # Step 3: Find JSON object/array
    # Look for first { or [
    start = text.find('{')
    if start == -1:
        start = text.find('[')
    if start == -1:
        raise ValueError("No JSON found in response")
    
    # Step 4: Find matching closing brace/bracket
    depth = 0
    in_string = False
    escape_next = False
    
    for i in range(start, len(text)):
        char = text[i]
        
        if escape_next:
            escape_next = False
            continue
        
        if char == '\\':
            escape_next = True
            continue
        
        if char == '"' and not escape_next:
            in_string = not in_string
            continue
        
        if in_string:
            continue
        
        if char == '{' or char == '[':
            depth += 1
        elif char == '}' or char == ']':
            depth -= 1
            if depth == 0:
                json_str = text[start:i+1]
                try:
                    return json.loads(json_str)
                except json.JSONDecodeError as e:
                    raise ValueError(f"Invalid JSON: {str(e)}")
    
    raise ValueError("Unclosed JSON structure")


def extract_json_simple(text: str) -> dict | list | None:
    """
    Simple JSON extraction that tries to parse the entire text.
    Falls back to extract_json if that fails.
    
    Args:
        text: Raw text that might contain JSON
    
    Returns:
        Parsed JSON or None if extraction fails
    """
    text = text.strip()
    
    # Try parsing the whole text first
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass
    
    # Fall back to full extraction
    try:
        return extract_json(text)
    except ValueError:
        return None

